### PR TITLE
Port from xml-rs to quick-xml

### PR DESCRIPTION
**Pull request by jcgruenhage**
Follow-up from #78. Things that happened since:
 - rebased this on the current changes on master
 - dropped the quick-xml impl into where the xml-rs one used to be, instead of offering quick-xml as a feature
 - fixed up the issue with pending collections
 - fixed a whole lot of clippy issues while I was at it

Thanks to @pr2502 for doing the heavy lifting here, most stuff was already in pretty good shape so I mostly just moved stuff around and did a few little touch-ups here and there.

### CODE REVIEW

File: rustfmt.toml [hunk 1]
```diff
@@ -1 +1 @@
-merge_imports = true
+imports_granularity="Crate"
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply rustfmt.toml:1]: rustfmt complained that merge_imports is deprecated and that I should use imports_granularity instead. It's in a separate commit together with all the clippy lints that have been fixed.

File: src/stream/binary_reader.rs [hunk 2]
```diff
@@ -418,7 +418,7 @@ mod tests {
     fn streaming_parser() {
         use crate::stream::Event::*;
 
-        let reader = File::open(&Path::new("./tests/data/binary.plist")).unwrap();
+        let reader = File::open(Path::new("./tests/data/binary.plist")).unwrap();
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/binary_reader.rs:421]: yep, there's a bunch of those, but those pre-existed. I just removed the referencing because clippy complained, but you're right of course, we don't need to create a Path here. I'll remove them.

File: src/dictionary.rs [hunk 3]
```diff
@@ -14,6 +14,7 @@ use std::{
 use crate::Value;
 
 /// Represents a plist dictionary type.
+#[derive(Default, Clone, PartialEq)]
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/dictionary.rs:17]: IndexMap does not provide an `Eq` implementation, so we can't derive `Eq` on it.

File: src/stream/xml_reader.rs [hunk 4]
```diff
@@ -1,221 +1,258 @@
...
+                xml_reader,
+                closed_element: None,
+                element_stack: Vec::new(),
+            },
         }
     }
+}
 
-    fn read_content(&mut self) -> Result<String, Error> {
+fn from_xml_error(err: XmlReaderError) -> ErrorKind {
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:54]: This is indeed an error conversion, I've changed it to use `From` instead.

File: src/serde_tests.rs [hunk 5]
```diff
@@ -813,6 +813,15 @@ fn dictionary_deserialize_dictionary_in_struct() {
     );
 }
 
+#[test]
+fn empty_array_and_dictionary() {
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/serde_tests.rs:817]: removed it

File: src/error.rs [hunk 6]
```diff
@@ -61,9 +61,10 @@ pub(crate) enum ErrorKind {
     Serde(String),
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone, Copy)]
 pub(crate) enum FilePosition {
-    LineColumn(u64, u64),
+    // Never used anymore?
+    //LineColumn(u64, u64),
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/error.rs]: Simplified to a tuple struct for now

File: src/stream/binary_reader.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/binary_reader.rs]: it's been done ^^

File: src/stream/xml_reader.rs [hunk 7]
```diff
@@ -1,221 +1,258 @@
...
+                        _ => return Err(self.with_pos(ErrorKind::UnknownXmlElement)),
+                    }
+                }
+                XmlEvent::End(name) => {
                     // Check the corrent element is being closed
                     match self.element_stack.pop() {
-                        Some(ref open_name) if &name.local_name == open_name => (),
-                        Some(ref _open_name) => {
+                        Some(open_name) if name.local_name() == open_name.as_ref() => {}
+                        Some(_open_name) => {
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:209]: Yep, makes sense, this looks a bit cleaner to me.

**Comment by jcgruenhage** [comment]
Thanks for your feedback, @RickyDaMa. I've applied most of your suggestions, they do improve the code quite a bit. I'll be honest: I haven't completely read the code that I committed, just the bits I needed to touch to apply the feedback from over at #78 :sweat_smile:, as well as a bit required for #52 later on, although I haven't started implementation on that one yet, as I'd want this here and #85 to be merged first.

**Comment by ebarnard** [comment]
@RickyDaMa Thanks for doing the review for me 😁! I'll try and take a look at it later today.

File: src/stream/binary_writer.rs [hunk 8]
```diff
@@ -527,7 +526,7 @@ impl<W: Write> BinaryWriter<W> {
                     self.writer.write_exact(&buf)?;
                 } else {
                     let mut buf: [_; 9] = [0x87, 0, 0, 0, 0, 0, 0, 0, 0];
-                    buf[1..].copy_from_slice(&(v as u64).to_be_bytes());
+                    buf[1..].copy_from_slice(&v.to_be_bytes());
```
Comment (ebarnard) [inline src/stream/binary_writer.rs:529]: Please keep this, or change to `u64::to_be_bytes(v)`. I want it to be very clear what the type of `v` is.

File: src/stream/binary_writer.rs [hunk 9]
```diff
@@ -713,24 +712,23 @@ mod tests {
 
         let streaming_parser = BinaryReader::new(Cursor::new(buf_inner));
 
-        let events: Vec<Result<_, _>> = streaming_parser.collect();
```
Comment (ebarnard) [inline src/stream/binary_writer.rs:716]: This collect into a `Vec` is intentional so that any errors that occur during reading are reported before any errors that occur during `Value::from_events`.

File: src/serde_tests.rs [hunk 10]
```diff
@@ -848,34 +848,64 @@ fn dictionary_serialize_xml() {
     let buf = buf.into_inner();
     let xml = std::str::from_utf8(&buf).unwrap();
 
-    let comparison = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
-<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">
-<plist version=\"1.0\">
+    let comparison = r#"<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ebarnard) [inline src/serde_tests.rs:851]: I've had problems in the past with text editors replacing tabs with spaces. I'd rather keep the escaped version, even though it doesn't look very nice.

File: src/stream/mod.rs [hunk 11]
```diff
@@ -96,9 +97,29 @@ impl XmlWriteOptions {
...
+    /// `XmlWriteOptions::with_indent`.
+    pub fn indent_string(self, indent_str: impl Into<Cow<'static, str>>) -> Self {
+        let indent_str = indent_str.into();
+        let indent_str = indent_str.as_ref();
+
+        if indent_str.is_empty() {
+            return self.with_indent(0, 0);
+        }
+
+        assert!(
```
Comment (ebarnard) [inline src/stream/mod.rs:112]: I'm not really sure what to do in this case - I can't imagine anyone is using a non-repeating indent string - so maybe the panic is ok.

File: src/stream/xml_reader.rs [hunk 12]
```diff
@@ -1,221 +1,262 @@
...
-
-    fn read_next(&mut self) -> Result<Option<OwnedEvent>, Error> {
+    fn read_next(&mut self, buffer: &mut Vec<u8>) -> Result<Option<OwnedEvent>, Error> {
         loop {
-            match self.next_event() {
-                Ok(XmlEvent::StartDocument { .. }) => {}
-                Ok(XmlEvent::StartElement { name, .. }) => {
+            if let Some(closed_name) = self.closed_element.take() {
+                // Check the corrent element is being closed
+                match self.element_stack.pop() {
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:124]: [`Reader::check_end_names`](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.check_end_names) looks like it might do this check for us.

File: src/stream/xml_reader.rs [hunk 13]
```diff
@@ -1,221 +1,262 @@
...
+                        }
+                        _ => return Err(self.with_pos(ErrorKind::UnknownXmlElement)),
+                    }
+                }
+                XmlEvent::Empty(name) => {
+                    match name.local_name() {
+                        b"true" => return Ok(Some(Event::Boolean(true))),
+                        b"false" => return Ok(Some(Event::Boolean(false))),
+
+                        b"array" | b"dict" => {
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:191]: It looks like using [Reader::expand_empty_elements](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.expand_empty_elements) can avoid us having to duplicate logic between `<array>...</array>` and `<array />` events.

File: src/stream/xml_reader.rs [hunk 14]
```diff
@@ -1,221 +1,262 @@
...
                 }
-
-                Ok(XmlEvent::Characters(c)) => {
-                    if !is_whitespace_str(&c) {
-                        return Err(
-                            self.with_pos(ErrorKind::UnexpectedXmlCharactersExpectedElement)
-                        );
-                    }
+                XmlEvent::Eof if self.element_stack.is_empty() => return Ok(None),
+                XmlEvent::Eof => return Err(self.with_pos(ErrorKind::UnclosedXmlElement)),
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:223]: Again, I think we get this check for free if `Reader::check_end_names` is enabled via [`Error::UnexpectedEof`](https://docs.rs/quick-xml/latest/quick_xml/enum.Error.html).

File: src/stream/xml_reader.rs [hunk 15]
```diff
@@ -1,221 +1,262 @@
...
             }
         }
     }
-
-    fn with_pos(&self, kind: ErrorKind) -> Error {
-        kind.with_position(convert_xml_pos(self.xml_reader.position()))
-    }
 }
 
 impl<R: Read> Iterator for XmlReader<R> {
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:239]: I know I'm being a bit picky, but can this go above the impl block for `ReaderState`.

File: src/stream/xml_reader.rs [hunk 16]
```diff
@@ -1,221 +1,262 @@
...
-                    return Ok("".to_owned());
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
+            match self.read_xml_event(buffer)? {
+                XmlEvent::Text(text) => {
+                    let pos = self.xml_reader_pos();
+                    let unespcaped = text
+                        .unescaped()
+                        .map_err(|err| ErrorKind::from(err).with_position(pos))?;
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:98]: Does `.map_err(|e| self.with_pos(ErrorKind::from(e)))?;` work here?

File: src/stream/xml_writer.rs [hunk 17]
```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
...
                 }
             }
+            this.pending_collection = None;
             this.expecting_key = this.stack.last() == Some(&Element::Dictionary);
             Ok(())
         })
     }
 
     fn write_boolean(&mut self, value: bool) -> Result<(), Error> {
+        self.handle_pending_collection()?;
```
Comment (ebarnard) [inline src/stream/xml_writer.rs:226]: Can this go in `write_value_event` so it's not duplicated across all the functions below?

File: src/stream/xml_writer.rs [hunk 18]
```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
...
+        self.pending_collection = Some(PendingCollection::Dictionary);
+        Ok(())
     }
 
     fn write_end_collection(&mut self) -> Result<(), Error> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
-                (Some(Element::Dictionary), true) => {
+            match (
```
Comment (ebarnard) [inline src/stream/xml_writer.rs:186]: This match is getting a bit hard to follow. I think you can just check beforehadn if there is a pending collection, and if so just write the empty tag and return before the existing match statement.

**Review by ebarnard — CHANGES_REQUESTED** [review]
It's looking pretty good @jcgruenhage, just a few changes. I've only reviewed the first commit, please could you drop the clippy changes from this PR for the time being and put them in another once this is merged.

File: src/date.rs [hunk 19]
```diff
@@ -35,7 +35,7 @@ impl Date {
     }
 
     /// Converts the `Date` to an XML plist date string.
-    pub fn to_xml_format(&self) -> String {
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/date.rs:38]: It's going to be dropped anyway, because it's part of the second commit. But yes, stable API, good argument, even though it is just 16 bytes and it does implement `Copy`, which is why clippy does suggest changing that.e

**Comment by jcgruenhage** [comment]
Thanks for the extensive review @ebarnard, I'll try to get the requested changes in today :)

File: src/stream/xml_reader.rs [hunk 20]
```diff
@@ -131,7 +131,7 @@ impl<R: Read> ReaderState<R> {
 
                 match closed_name.as_ref() {
                     b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
-                    b"plist" | _ => {}
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/xml_reader.rs]: Part of the second commit, so this will be dropped out of the PR.

File: src/stream/xml_reader.rs [hunk 21]
```diff
@@ -216,7 +216,7 @@ impl<R: Read> ReaderState<R> {
 
                     match name.local_name() {
                         b"array" | b"dict" => return Ok(Some(Event::EndCollection)),
-                        b"plist" | _ => (),
```
[the remark this answers is not in the document]
Comment (jcgruenhage) [reply src/stream/xml_reader.rs]: Part of the second commit, so this will be dropped out of the PR.

File: src/stream/binary_writer.rs (hunk 8 above)
Comment (jcgruenhage) [reply src/stream/binary_writer.rs:529]: Part of the second commit, so this will be dropped out of the PR.

File: src/stream/binary_writer.rs (hunk 9 above)
Comment (jcgruenhage) [reply src/stream/binary_writer.rs:716]: Part of the second commit, so this will be dropped out of the PR.

File: src/stream/mod.rs (hunk 11 above)
Comment (jcgruenhage) [reply src/stream/mod.rs:112]: Considering this is a public function, this is technically a breaking change for those who use a non-repeating indent string. I'd still argue that keeping it this way and having it panic in that case is still better than making it a breaking change for everyone.

I've added a doc string for `with_indent`, which I seem to have forgotten in my first round of refactoring, and added a deprecation warning for indent_string, to reduce the chance of people getting bitten by this.

File: src/stream/xml_reader.rs (hunk 12 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:124]: oh, and it's set to `true` by default, so we can skip this check completely. Good catch!

Comment (jcgruenhage) [reply src/stream/xml_reader.rs:124]: I think that means we can get rid of everything related to `closed_element` and `element_stack`, which would reduce the code complexity quite a bit :)

File: src/stream/xml_writer.rs (hunk 17 above)
Comment (jcgruenhage) [reply src/stream/xml_writer.rs:226]: That needed a bit of restructuring, but yeah, good call.

File: src/stream/xml_writer.rs (hunk 18 above)
Comment (jcgruenhage) [reply src/stream/xml_writer.rs:186]: split it up into two matches now

File: src/stream/xml_reader.rs (hunk 16 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs]: yep, it does!

File: src/serde_tests.rs (hunk 10 above)
Comment (jcgruenhage) [reply src/serde_tests.rs]: converted it back, seems reasonable to me

**Comment by jcgruenhage** [comment]
I've addressed all of the feedback except for the `Reader::check_end_names` thing so far, because that'll be a slightly bigger change. I hope I'll get to that later as well though.

File: src/stream/xml_reader.rs (hunk 12 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:123]: Yep, it does mean that, I've implemented it now.

File: src/stream/xml_reader.rs (hunk 13 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:190]: :+1:

File: src/stream/xml_reader.rs (hunk 15 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:238]: Moved it up.

**Comment by jcgruenhage** [comment]
Okay, @ebarnard, your feedback should now be fully addressed, let me know if there is anything else you'd want changed here.

**Comment by pr2502** [comment]
hi @jcgruenhage thank you for finishing this! i'm happy you decided to tag me as a co-author, but may i ask you to replace my old tag with `max <36980911+pr2502@users.noreply.github.com>`. thanks a lot ^^

**Comment by jcgruenhage** [comment]
Sure thing!

File: src/stream/xml_reader.rs [hunk 22]
```diff
@@ -1,219 +1,205 @@
...
-                }
-                Err(err) => return Err(from_xml_error(err)),
+impl From<XmlReaderError> for ErrorKind {
+    fn from(err: XmlReaderError) -> Self {
+        match err {
+            XmlReaderError::Io(err) if err.kind() == io::ErrorKind::UnexpectedEof => {
+                ErrorKind::UnexpectedEof
             }
+            XmlReaderError::Io(err) => {
+                let err = if let Some(code) = err.raw_os_error() {
```
Comment (ebarnard) [inline src/stream/xml_reader.rs:55]: Can this not just be `XmlReaderError::Io(err) => ErrorKind::Io(err)`?

File: src/stream/xml_writer.rs [hunk 23]
```diff
@@ -157,28 +145,65 @@ impl<W: Write> XmlWriter<W> {
...
     }
 
     fn write_end_collection(&mut self) -> Result<(), Error> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
+            match this.pending_collection.take() {
+                Some(PendingCollection::Array) => {
+                    this.xml_writer
+                        .write_event(XmlEvent::Empty(BytesStart::borrowed_name(b"array")))?;
+                    this.expecting_key = this.stack.last() == Some(&Element::Dictionary);
```
Comment (ebarnard) [inline src/stream/xml_writer.rs:192]: I don't think this, or the copy below is needed. `pending_collection` will only be `Some` if this is called straight after `write_start_dictionary` or `write_start_array` in which case the value of `expecting_key` will have remained unchanged so the existing value is always valid.

**Review by ebarnard — COMMENTED** [review]
Two more small questions and then we're good to go. Thank you so much @jcgruenhage for taking time to work on this.

File: src/stream/xml_writer.rs (hunk 23 above)
Comment (ebarnard) [reply src/stream/xml_writer.rs:192]: No need to change this now by the way.

File: src/stream/xml_writer.rs [hunk 24]
```diff
@@ -30,8 +27,12 @@ pub struct XmlWriter<W: Write> {
...
-    empty_namespace: Namespace,
+    pending_collection: Option<PendingCollection>,
+}
+
+enum PendingCollection {
+    Array,
+    Dictionary,
 }
 
 impl<W: Write> XmlWriter<W> {
```
Comment (ebarnard) [inline src/stream/xml_writer.rs:38]: If you are changing anything else, please can you correct my type of `fearure` below!

File: src/stream/xml_reader.rs (hunk 22 above)
Comment (jcgruenhage) [reply src/stream/xml_reader.rs:55]: apparently it can be, yeah. Some of the code that I didn't touch myself yet ^^

File: src/stream/xml_writer.rs (hunk 23 above)
Comment (jcgruenhage) [reply src/stream/xml_writer.rs:192]: I'd need to look into what is happening here in more detail to understand why it is how it is, but with those statements removed the tests fail.

File: src/stream/xml_writer.rs (hunk 24 above)
Comment (jcgruenhage) [reply src/stream/xml_writer.rs:38]: I'll include it as a separate commit :)

**Comment by jcgruenhage** [comment]
@ebarnard I think this is ready now. One question: What's going to be the next released version? Asking because of the deprecation warning on `XmlWriteOptions::indent_string`

**Comment by ebarnard** [comment]
Think it should be 1.4

**Review by ebarnard — APPROVED** [review]

**Comment by jcgruenhage** [comment]
Then the deprecation is already good as it is. With your approval on it now, do you want to merge it?

**Comment by lucasfernog** [comment]
@ebarnard do you have an idea when this could be released?

**Comment by ebarnard** [comment]
v1.4 has now been released.
